### ISSUE TITLE
Apollo federation _service.sdl field returns invalid schema missing scalar types

### ISSUE THREAD

**Issue by kdawgwilk** [hasura/graphql-engine#9115]
### Version Information

Server Version: 2.13.0

### Environment

CE edition

```
HASURA_GRAPHQL_EXPERIMENTAL_FEATURES=apollo_federation
```

### What is the current behaviour?

Running
```gql
query {
  _service {
    sdl
  }
}
```

Returns an invalid schema that is missing scalars such as `uuid`

### What is the expected behaviour?

The following query returns a fully valid SDL schema that passes official GraphQL spec requirements such as defining the scalar types being used in other fields
```gql
query {
  _service {
    sdl
  }
}
```

### How to reproduce the issue?

1. Spin up hasura instance with the experimental apollo_federation feature enabled
```
HASURA_GRAPHQL_EXPERIMENTAL_FEATURES=apollo_federation
```
2. Run the following query
```gql
query {
  _service {
    sdl
  }
}
```
4. Search generated sdl for `scalar uuid`

### Please provide any traces or logs that could help here.

```
/usr/app/node_modules/@apollo/gateway/src/supergraphManagers/IntrospectAndCompose/index.ts:118
      throw Error(
            ^
Error: A valid schema couldn't be composed. The following composition errors were found:
	[hasura] Unknown type uuid
```

### Any possible solutions?

These scalars do seem to be present when introspection the graph so there must be a difference between introspection and producing this sdl and the bug likely is due to the difference in functionality between those two

### If the bug is confirmed, would you be willing to submit a PR?

I wish I knew enough Haskell to be able to do this 😞 

### Keywords

apollo, federation, subgraph, IntrospectAndCompose, Unknown type uuid

**Comment by AienTech** [comment]
just wanted to mention that this one is quite a crucial part of using hasura in a gateway. hope to see an update in this regard soon.

**Comment by paritosh-08** [comment]
Thanks for raising the issue. We are actively working on a fix. @kdawgwilk @AienTech, can you please try out the docker image `hasurabuild/graphql-engine:mono-pr-6541-scal.ubuntu` and let us know if the issue is fixed?

**Comment by AienTech** [comment]
@paritosh-08 just tested it and it looks so far fine. tnx

**Comment by devakrishna33** [comment]
Facing the same issue, tested with `hasurabuild/graphql-engine:mono-pr-6541-scal.ubuntu` it works fine, thanks for that. Is it fine to use this build in production until a version is released with this fix?

**Comment by kdawgwilk** [comment]
I was unable to test the fix due to the `hasurabuild/graphql-engine:mono-pr-6541-scal.ubuntu` image is only built for amd64 CPU arch and I am running on an M1 arm Mac.  It would be great to see those nightly images built for both arch like the `hasura/graphql-engine` are

**Comment by devakrishna33** [comment]
For anyone else facing this issue, here's a temporary workaround.

`
const { printSchema } = require("graphql/utilities");
import { introspectSchema } from "@graphql-tools/wrap";
const { print } = require("graphql");
import fetch from "cross-fetch";
import {
    ApolloGateway,
    GraphQLDataSourceProcessOptions,
    RemoteGraphQLDataSource,
} from "@apollo/gateway";
import { GatewayGraphQLRequestContext } from "@apollo/server-gateway-interface";

const executor = async ({ document, variables, context }: any) => {
    const query = print(document);

    const fetchResult = await fetch(process.env.HASURA_ENDPOINT!, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "X-Hasura-User-ID": context?.userId,
            "X-Hasura-Admin-Secret": process.env.HASURA_GRAPHQL_ADMIN_SECRET!,
            "X-Hasura-Role": "admin",
        },
        body: JSON.stringify({ query, variables }),
    });
    return fetchResult.json();
};

const getHasuraSDL = async () => {
    const hasuraSchema = await introspectSchema(executor);

    // retunr the SDL for the hasura schema
    return printSchema(hasuraSchema);
};


class HasuraDataSource extends RemoteGraphQLDataSource {
    willSendRequest({ request, ...rest }: GraphQLDataSourceProcessOptions) {
        if (request.http)
            request.http.headers.set(
                "X-Hasura-Admin-Secret",
                process.env.HASURA_GRAPHQL_ADMIN_SECRET!
            );
    }

    async didReceiveResponse({
        response,
        request,
    }: Required<
        Pick<GatewayGraphQLRequestContext, "request" | "response" | "context">
    >) {
        if (
            request.query ===
            "query __ApolloGetServiceDefinition__ { _service { sdl } }"
        ) {
            // main solution happens here
            // manually fetch the hasura schema and return it as an SDL
            const hasuraSDL = await getHasuraSDL();
            response.data = {
                _service: {
                    sdl: hasuraSDL,
                },
            };
        }

        // Return the response, even when unchanged.
        return response;
    }
}

const gateway = new ApolloGateway({
    serviceList: [
        {
            name: "hasura",
            url: process.env.HASURA_ENDPOINT,
        },
    ],
    introspectionHeaders: {
        "X-Hasura-Admin-Secret": process.env.HASURA_GRAPHQL_ADMIN_SECRET!,
        "X-Hasura-Role": "admin",
    },
    buildService({ url }) {
        return new HasuraDataSource({ url });
    },
});
`

**Comment by rahulagarwal13** [comment]
Hey @kdawgwilk - regret that you faced that problem. We will look into supporting other architectures for temporary builds but unlikely to happen in the short term. The fix has been merged in the upcoming `v2.16.0-beta.1` release (latest by the end of this week). Hope you can test the fix in that.

Hey @devakrishna33 - no `hasurabuild/graphql-engine:mono-pr-6541-scal.ubuntu` is not recommended for production. We suggest waiting one week or so for the v2.16 to become stable for you to use the fix in production.

**Comment by rahulagarwal13** [comment]
Closing this issue out since got confirmation from @kdawgwilk that the issue is resolved for them in v2.16.
@devakrishna33  - v2.16 is out so you can use the fix in production now

**Comment by rahulagarwal13** [comment]
Closing this issue out since got confirmation from @kdawgwilk that the issue is resolved for them in v2.16.
@devakrishna33 - v2.16 is out so you can use the fix in production now
